deduplication-document-spark/deduplication-document-spark-impl/src/main/scala/pl/edu/icm/coansys/document/deduplication/DeduplicateDocuments.scala

```diff
@@ -0,0 +1,420 @@
+    val initialClusterCount = clustersToDeduplicate.count
+    //TODO: some statistics here on cluster, would be useful.
+
+    val tiledTasks = clustersToDeduplicate.flatMap(p => CartesianTaskSplit.parallelizeCluster(p._1, p._2, cfg.tileSize))
+    
+    tiledTasks.persist
+    val tileCount = tiledTasks.count;
+
+    println(f"Prepared $initialClusterCount clusters, and then split it to $tileCount tiles")
+    
+    //build (clusterId, Seq(docId)) rdd:    
+    val partialEqualityClusters = tiledTasks.flatMap(
+      task => {
+        val t0 = java.lang.System.currentTimeMillis()
+        val comparator = buildWorkComparator
+        val res = task.processPairs((a: DocumentWrapper, b: DocumentWrapper) =>
+          comparator.isDuplicate(a.getDocumentMetadata, b.getDocumentMetadata, null))
+        val time = (java.lang.System.currentTimeMillis() - t0) / 1000.0
+        //useful for identification of possible problem.
+        log.info(f"Finishing tile task ${task.taskId} in $time%.4f sec")
+        res.map((task.clusterId, _))
+      }
+    )
+    
+    val finalClusters = partialEqualityClusters.mapValues(List(_)).
+      reduceByKey(_ ++ _). //one long list of lists of ids for each cluster
+      map(pair => {
+        val t0 = java.lang.System.currentTimeMillis()
+        val res = CartesianTaskSplit.coalesceClusters(pair._2.asJava)
+        val tt = System.currentTimeMillis() - t0;
+        val clusterSize = pair._2.size
+        log.info(f"Finished tile coalesce task. (Cluster,time[s], size): ${pair._1}, ${tt / 1000.0}, ${pair._2.size}")
+        (pair._1, res)
+      }).
+      flatMap( //build proper ids for equality clusters.
+        p => {
+          val cid = p._1
+          val cl = p._2
+          cl.zipWithIndex.map(q => (cid + f"_${q._2}%03d", q._1))
+        }
+      )
+      
+    //now we got all the items in place    
+    finalClusters.persist
+    
+    printTopClusters(finalClusters, 100)
+
+    //count clusters, documents in clusters and number of documents to be deduced:
+    val finalClusterCount = finalClusters.count
+    val documentInFinalClusterCount = finalClusters.map(_._2.size).fold(0)(_ + _)
+    val documentRemovedDuringClusteringCount = documentInFinalClusterCount - finalClusterCount
+    println(f"Finally created $finalClusterCount clusters, containing $documentInFinalClusterCount documents, $documentRemovedDuringClusteringCount documents will be removed.")
+
+    // merge documents
+    val docIdWithClusterId = finalClusters.flatMapValues(x => x).
+      map(v => (v._2, v._1))
+    val documentWrappersPrepared = wrappers.leftOuterJoin(docIdWithClusterId);
+    val mergedDocuments = documentWrappersPrepared.filter(_._2._2.isDefined).
+      map(x => (x._2._2, List(x._2._1))).foldByKey(List())(_ ++ _). //get lists of cluster documents by cluster id
+      map(kv => {
+        val doc = mergeDocuments(kv._2)
+        (doc.getDocumentMetadata.getKey, doc)
+      })
```

Comment:
The pipeline could be split into steps for more clarity.

After all changes listed above the code would look like:

```scala
    val clusterIdWithDocument: RDD[(Option[String], DocumentWrapper)] =
      documentWrappersPrepared
        .values
        .filter(_._2.isDefined)
        // changed order of joining would allow to remove 'swap' operation here
        .map(_.swap)
    val clusteredDocuments: RDD[Iterable[DocumentWrapper]] =
      clusterIdWithDocument
        .groupByKey()
        .values
    val mergedDocuments: RDD[(String, DocumentWrapper)] =
      clusteredDocuments
        .map(mergeDocuments)
        .keyBy(_.getDocumentMetadata.getKey)